### PR TITLE
Fix TODO in pkg/controller/nodecontroller.go line 472

### PR DESCRIPTION
**Pull request by WeixuZhuang**
The code now calculates and find out the CIDRs for every node in any sync period.
I will fix this TODO by maintaining a set for available CIDRs left. Firstly, I will
insert 256 CIDRs into the available set. Once someone get one CIDR, remove this CIDR
from the available set. If one node get deleted, we will reinsert the CIDR associates
with this node back to available CIDR. Once there are nothing left in available CIDR set,
generate another 256 CIDRs and insert them into the available set. As a result, we do not
need to generate CIDRs in every monitor process and we only need to assign CIDR to node
which does not have it.

This commit also fix the error that CIDR may overflow when we use the function
generateCIDRs. There will be no more ip overflowing, all assigan CIDR will be valid

### CODE REVIEW

**Comment by zhengguoyong** [comment]
run "go fmt pkg/controller/node/nodecontroller.go" before commit will help resolve the build failed.
:)

**Comment by WeixuZhuang** [comment]
@zhengguoyong , thank you for your help. I have uploaded new commit, @gmarek , could you please review my codes when you have time?

Thanks

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -193,12 +202,32 @@ func NewNodeController(
 	return nc
 }
 
+// Generate new Available CIDRs index and insert them into availableCIDRs set
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: godoc convention would be:

"generateNewAvailableCIDRs does xyz"

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -274,6 +303,20 @@ func generateCIDRs(clusterCIDR *net.IPNet, num int) sets.String {
 	return res
 }
 
+// Translate pod CIDRs index to pod CIDRs that could be assigned to node.
+// It will also check for overflow which make sure CIDR is valid
+func translateCIDRs(clusterCIDR *net.IPNet, num int) string {
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: godoc style: translateCIDRs does

File: pkg/controller/node/nodecontroller.go [hunk 3]
```diff
@@ -343,6 +386,16 @@ func forcefullyDeletePod(c client.Interface, pod *api.Pod) {
 	}
 }
 
+// Translate CIDR back to CIDR index and insert this index back to availableCIDRs set
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: releaseCIDR does

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -467,24 +527,21 @@ func (nc *NodeController) monitorNodeStatus() error {
 
 // reconcileNodeCIDRs looks at each node and assigns it a valid CIDR
 // if it doesn't currently have one.
+// Examinate the availableCIDRs set first, if no more CIDR in it, generate more.
```
Comment (derekwaynecarr) [inline pkg/controller/node/nodecontroller.go]: Examines the ...

**Comment by WeixuZhuang** [comment]
Thanks @derekwaynecarr , I have updated the documents

File: pkg/controller/node/nodecontroller.go [hunk 5]
```diff
@@ -112,6 +113,11 @@ type NodeController struct {
 	nodeStore      cache.StoreToNodeLister
 
 	forcefullyDeletePod func(*api.Pod)
+	availableCIDRs      sets.Int
+	// Record the first CIDR which has not been put into the availableCIDR set
+	firstAvailable int
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: firstNotAvailable

File: pkg/controller/node/nodecontroller.go [hunk 6]
```diff
@@ -193,12 +202,32 @@ func NewNodeController(
 	return nc
 }
 
+// generateNewAvailabeCIDRs does generating new Available CIDRs index and insert
+// them into availableCIDRs set. Everytime it will generate 256 new CIDRs,
+// once there are no more CIDRs in the network, return false
+func (nc *NodeController) generateNewAvailabeCIDRs() bool {
+	if nc.firstAvailable+256 > nc.maxCIDRs {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Is there a reason why we don't want to return sets smaller than 256?

File: pkg/controller/node/nodecontroller.go [hunk 7]
```diff
@@ -193,12 +202,32 @@ func NewNodeController(
...
+		glog.V(4).Infof("CIDR is not enough")
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: "Out of CIDRs" is probably a better wording. Plus I think it should be a Warning.

File: pkg/controller/node/nodecontroller.go [hunk 8]
```diff
@@ -193,12 +202,32 @@ func NewNodeController(
...
+	return true
+}
+
 // Run starts an asynchronous loop that monitors the status of cluster nodes.
 func (nc *NodeController) Run(period time.Duration) {
 	go nc.nodeController.Run(util.NeverStop)
 	go nc.podController.Run(util.NeverStop)
 	// Incorporate the results of node status pushed from kubelet to master.
 	go util.Until(func() {
+		// initiate all CIDRs as available here
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Why are you doing this in the loop? You're basically regenerating available CIDRs in each iteration.

File: pkg/controller/node/nodecontroller.go [hunk 9]
```diff
@@ -274,6 +303,20 @@ func generateCIDRs(clusterCIDR *net.IPNet, num int) sets.String {
...
+// assigned to node. It will also check for overflow which make sure CIDR is valid
+func translateCIDRs(clusterCIDR *net.IPNet, num int) string {
+	cidrIP := clusterCIDR.IP.To4()
+	b1 := (num >> 8) + int(cidrIP[1])
+	b2 := (num % 256) + int(cidrIP[2])
+	if b2 > 255 {
+		b2 = b2 % 256
+		b1 = b1 + 1
+	}
+	res := fmt.Sprintf("%d.%d.%d.0/24", cidrIP[0], b1, b2)
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Add the same TODO as in generateCIDRs: `// TODO: Make the CIDRs configurable.`

Also delete generateCIDRs, as it should be unused now.

File: pkg/controller/node/nodecontroller.go [hunk 10]
```diff
@@ -274,6 +303,20 @@ func generateCIDRs(clusterCIDR *net.IPNet, num int) sets.String {
 	return res
 }
 
+// translateCIDRs does translating pod CIDRs index to pod CIDRs that could be
+// assigned to node. It will also check for overflow which make sure CIDR is valid
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: It doesn't check right now. Either implement it or remove the comment.

File: pkg/controller/node/nodecontroller.go [hunk 11]
```diff
@@ -274,6 +303,20 @@ func generateCIDRs(clusterCIDR *net.IPNet, num int) sets.String {
...
+func translateCIDRs(clusterCIDR *net.IPNet, num int) string {
+	cidrIP := clusterCIDR.IP.To4()
+	b1 := (num >> 8) + int(cidrIP[1])
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: why not `* 256`?

File: pkg/controller/node/nodecontroller.go [hunk 12]
```diff
@@ -343,6 +386,17 @@ func forcefullyDeletePod(c client.Interface, pod *api.Pod) {
 	}
 }
 
+// releaseCIDR does translating CIDR back to CIDR index and insert this index
+// back to availableCIDRs set
+func (nc *NodeController) releaseCIDR(freeCIDR string) {
+	CIDRsArray := strings.Split(freeCIDR, ".")
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: If I'm not mistaken it's OK for /24, but it won't work for /26 or sth, as last token will have /26 at the end. It's not implemented now, but it's a straightforward change here, so let's make it future proof.

File: pkg/controller/node/nodecontroller.go [hunk 13]
```diff
@@ -343,6 +386,17 @@ func forcefullyDeletePod(c client.Interface, pod *api.Pod) {
...
+	cidrIP := nc.clusterCIDR.IP.To4()
+	CIDRsIndexOne, _ := strconv.Atoi(CIDRsArray[1])
+	CIDRsIndexTwo, _ := strconv.Atoi(CIDRsArray[2])
+	release := (CIDRsIndexOne-int(cidrIP[1]))*256 + CIDRsIndexTwo - int(cidrIP[2])
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: What will happen here if overflow occurred in `translateCIDRs`?

File: pkg/controller/node/nodecontroller.go [hunk 14]
```diff
@@ -414,7 +468,7 @@ func (nc *NodeController) monitorNodeStatus() error {
 				}
 			}
 			if lastReadyCondition.Status == api.ConditionUnknown &&
-				decisionTimestamp.After(nc.nodeStatusMap[node.Name].probeTimestamp.Add(nc.podEvictionTimeout)) {
+				decisionTimestamp.After(nc.nodeStatusMap[node.Name].probeTimestamp.Add(nc.podEvictionTimeout-gracePeriod)) {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Please revert this change.

File: pkg/controller/node/nodecontroller.go [hunk 15]
```diff
@@ -453,11 +507,18 @@ func (nc *NodeController) monitorNodeStatus() error {
...
 						continue
 					}
-
-					if err := nc.kubeClient.Nodes().Delete(node.Name); err != nil {
+					freeCIDR := ""
+					if node.Spec.PodCIDR != "" {
+						freeCIDR = node.Spec.PodCIDR
+					}
+					err = nc.kubeClient.Nodes().Delete(node.Name)
+					if err != nil {
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Please revert the split - we're using the form `if err := ...; err != nil {` everywhere.

File: pkg/controller/node/nodecontroller.go [hunk 16]
```diff
@@ -453,11 +507,18 @@ func (nc *NodeController) monitorNodeStatus() error {
 						nc.evictPods(node.Name)
 						continue
 					}
-
-					if err := nc.kubeClient.Nodes().Delete(node.Name); err != nil {
+					freeCIDR := ""
+					if node.Spec.PodCIDR != "" {
+						freeCIDR = node.Spec.PodCIDR
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: It's not `freeCIDR` it's an `assignedCIDR` or sth like that.

File: pkg/controller/node/nodecontroller.go (hunk 12 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: This is designed for /24, if we change it to /26 or make it configurable, we need to change this function a little bit. It will not be a big change if we change /24 to /x. This has the same TODO as TODO in generateCIDRs: // TODO: Make the CIDRs configurable.

File: pkg/controller/node/nodecontroller.go (hunk 5 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: This var defines the first CIDR which has not been put into the availableCIDRs set. For example, we have totally 300 nodes. They need totally 300 CIDRs. Since the node controller will generate 256 new available CIDRs each time generateNewAvailabeCIDRs() being called, the function generateNewAvailabeCIDRs() will be called twice. Now the firstAvailable will be 512 which means next time when we want to generate new CIDRs and insert it into availableCIDRs set(call function generateNewAvailabeCIDRs()). Next time we we call generateNewAvailabeCIDRs(), it should begin from No. 512th CIDR which mean the firstAvailable=512.

File: pkg/controller/node/nodecontroller.go (hunk 6 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: This is checking if we do not have any available CIDRs which has not being generated(or inserted into availbleCIDRs set). maxCIDRs represent the maximum number of CIDRs we have in our system(based on clusterCIDR *net.IPNet). Checking nc.firstAvailable+256 > nc.maxCIDRs is because everytime we call generateNewAvailabeCIDRs(), we will generate 256 CIDRs, we want to make sure they are enough CIDRs have not been generated yet.

File: pkg/controller/node/nodecontroller.go (hunk 13 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: It will not happen since translateCIDRs only translate index of CIDR into real CIDR. The index overflow(means no more CIDR in the system) checking is in generateNewAvailabeCIDRs(), and real CIDR overflow(means the CIDR is not valid CIDR, the example of not valid is 10.2.257.0/16) checking is in translateCIDRs.

**Comment by WeixuZhuang** [comment]
@gmarek , thank you so much for your reviewing. Sorry for late reply due to the Thanksgiving vacation. I have update the codes based on your review. Thank you.

File: pkg/controller/node/nodecontroller.go (hunk 5 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Yup, I understand what it does. I only mean that it should be called firstNotAvailable, firstUnassigned or something along those lines. firstAvailable made me think it's first CIDR to be assign to a node, which is not the case.

File: pkg/controller/node/nodecontroller.go (hunk 6 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Yup. I'm only asking why if you have, say, 64 CIDRs left you claim that you're out of them?

File: pkg/controller/node/nodecontroller.go (hunk 12 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I'm still not a fan of this logic - can you please strip the '/x' from the end?

File: pkg/controller/node/nodecontroller.go [hunk 17]
```diff
@@ -261,16 +290,18 @@ func (nc *NodeController) Run(period time.Duration) {
 	}, nodeEvictionPeriod, util.NeverStop)
 }
 
-// Generates num pod CIDRs that could be assigned to nodes.
-func generateCIDRs(clusterCIDR *net.IPNet, num int) sets.String {
-	res := sets.NewString()
+// translateCIDRs does translating pod CIDRs index to pod CIDRs that could be
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: translateCIDRs translates pod CIDR index to the CIDR that could be assigned to the node.

File: pkg/controller/node/nodecontroller.go (hunk 13 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: OK, I did some algebraic exercises and it looks good.

File: pkg/controller/node/nodecontroller.go [hunk 18]
```diff
@@ -193,11 +202,31 @@ func NewNodeController(
...
+}
+
 // Run starts an asynchronous loop that monitors the status of cluster nodes.
 func (nc *NodeController) Run(period time.Duration) {
 	go nc.nodeController.Run(util.NeverStop)
 	go nc.podController.Run(util.NeverStop)
 	// Incorporate the results of node status pushed from kubelet to master.
+	// generate first 256 available CIDRs here
+	cidrIP := nc.clusterCIDR.IP.To4()
+	nc.firstAvailable = 0
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: This line is not needed - it'll be initialized with the 0 value.

File: pkg/controller/node/nodecontroller.go [hunk 19]
```diff
@@ -193,11 +202,31 @@ func NewNodeController(
...
+	nc.maxCIDRs = (256-int(cidrIP[1]))*256 - int(cidrIP[2])
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: This should be moved to the NodeController constructor.

File: pkg/controller/node/nodecontroller.go [hunk 20]
```diff
@@ -193,11 +202,31 @@ func NewNodeController(
...
+	nc.generateNewAvailabeCIDRs()
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: This line is not needed - when they're not available `reconcileNodeCIDRs` will generate them.

**Comment by gmarek** [comment]
This part of the change is starting to look good. You need to add some way to handle crashes of ControllerManager, when all the state from memory (i.e. availableCIDRs) will be gone and will need to be recreated.

File: pkg/controller/node/nodecontroller.go (hunk 6 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: I mean you are right in this case. Since we do not have much chance to reach using that many CIDRs, I just let some CIDRs being wasted

**Comment by WeixuZhuang** [comment]
working on it, will update the code soon

**Comment by WeixuZhuang** [comment]
my code now will recreate availableCIDRs set based on the nodes and pods in the cluster everytime we restart the node controller. It will work when the controller crash and the controller restarted

File: pkg/controller/node/nodecontroller.go (hunk 5 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Please change.

File: pkg/controller/node/nodecontroller.go (hunk 6 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Is there a good reason to do so?

File: pkg/controller/node/nodecontroller.go (hunk 17 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Please fix

File: pkg/controller/node/nodecontroller.go [hunk 21]
```diff
@@ -460,11 +509,17 @@ func (nc *NodeController) monitorNodeStatus() error {
 						nc.evictPods(node.Name)
 						continue
 					}
-
-					if err := nc.kubeClient.Nodes().Delete(node.Name); err != nil {
+					assignedCIDR := ""
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Drop if, just `assignedCIDR := node.Spec.PodCIDR`, it's cleaner.

File: pkg/controller/node/nodecontroller.go [hunk 22]
```diff
@@ -474,28 +529,60 @@ func (nc *NodeController) monitorNodeStatus() error {
...
-				continue
+	// check if the this node controller is restarted because of crash
+	// this will only be ran once when the controller being restarted(because of crashed) or newly start
+	if nc.needSync {
+		// if it is crashed, restore the availableCIDRs by generating CIDRs, insert them into availableCIDRs set
+		// and delete assigned CIDRs from the the availableCIDRs set
+		for _, node := range nodes.Items {
+			if node.Spec.PodCIDR != "" {
+				if nc.availableCIDRs.Has(nc.translateCIDRtoIndex(node.Spec.PodCIDR)) {
+					nc.availableCIDRs.Delete(nc.translateCIDRtoIndex(node.Spec.PodCIDR))
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: What if two Nodes have the same CIDR for some reason (e.g. one Node got disconnected for an hour but came back)? We'll generate all CIDRs and fall back to reassigning CIDR to one of them. As we're generating CIDRs in order can we check if we're in the range of 'known' ones?

File: pkg/controller/node/nodecontroller.go [hunk 23]
```diff
@@ -474,28 +529,60 @@ func (nc *NodeController) monitorNodeStatus() error {
...
+				} else {
+					for true {
+						result := nc.generateNewAvailabeCIDRs()
+						// CIDRs has been used up and still cannot find the node's CIDR
+						// this means the previous assigned CIDR to this node has problem, empty its CIDR so it can be reassigned
+						if !result {
+							nc.recordNodeStatusChange(&node, "CIDRNotAvailable")
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: I think we need to evict all Pods from this Node in such case @davidopp. We were not handling this correctly before though.

**Comment by gmarek** [comment]
Sorry for slow reply - I had to dig myself of other stuff:/ One more thought - we currently have 3072 CIDRs in default setup (10.244.0.0/16 cluster IP range). The code would be simpler if instead of generating CIDRs piece by piece, we'd have all 3072 of them from the start (e.g. kept in array of bools or sth similar). It'll use much less than 1MB of memory so we should be able to live with that.

File: pkg/controller/node/nodecontroller.go (hunk 22 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: Sorry, I am a little bit confused about the case you described. Do you mean if one Node got disconnected when we running the nc.needSync code(above code), its CIDR will not be deleted from the availableCIDRs set and may be reassign to another Node? If this is the case, it will be a little bit complicated. You are right, if we simply check the CIDR range of known nodes, it could be solved. But it will waste a lot of CIDRs for example, we have 1000 nodes at first and delete the first 999 nodes and then restart the node_controller. In this case, the first 999 CIDRs will be wasted and we will start assigning from 1001st CIDR.

The other solution will be maintaining a bool array to mark if that CIDR has been used. As you comment below, we currently has 3072 CIDRs in default setup and it will be ok to maintain a bool array(since we will use less than 1MB memory). It will both simplify the code and solve this problem, I will update the code in this way. Thanks

File: pkg/controller/node/nodecontroller.go (hunk 23 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: @gmarek , I don't quiet understand why we need to evict all Pods from this Node when the node controller restart. I did not find out what Pod information storing in node controller need to be restored when the node controller restart so that we need to evict all Pods and spawn new Pods.
@davidopp , what do you think? should we evict all Pods from all assigned CIDR Nodes in case(the node controller restart)?

Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Not when it's restarted, but when CIDR is changed. Pods running on the Node gets IPs from the CIDR assigned to a Node.

File: pkg/controller/node/nodecontroller.go (hunk 22 above)
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Sorry, you're right. I was thinking about the scenario when Node becomes unresponsive and comes back again. But either we'll delete Node object for it, and everything will need to be recreated from the start, or it won't be deleted, and nothing else will get CIDR assigned to it.

File: pkg/controller/node/nodecontroller.go (hunk 23 above)
Comment (WeixuZhuang) [inline pkg/controller/node/nodecontroller.go]: Oh, the previous code does not handle this case but my code will not have this problem since it will not change CIDRs each time

**Comment by WeixuZhuang** [comment]
@gmarek , I have updated the code based on your suggestions. I will generate all available CIDRs and insert them into availableCIDRs set when we start the node controller. As a result, we don't need to generate CIDR piece by piece. The code has been simplified.

Meanwhile, I did not evict all Pods in the node when I change the node CIDR. In my code, there is only one case when we will change the node CIDR. It is when we found the node CIDR is invalid(either out of range or being used by other nodes already). In this case, I will clean the CIDR of this node and reassign new CIDR to it. If you decide that we need to evict all Pods from this node(I agree with your opinion), I will make this change

Thanks for your help.

File: pkg/controller/node/nodecontroller.go [hunk 24]
```diff
@@ -111,6 +113,11 @@ type NodeController struct {
 	nodeStore      cache.StoreToNodeLister
 
 	forcefullyDeletePod func(*api.Pod)
+	availableCIDRs      sets.Int
```
Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: can you make this []bool?

Comment (gmarek) [inline pkg/controller/node/nodecontroller.go]: Or let's forget about this comment. It's probably not worth the effort. Hash maps are generally expensive, AFAIK go implementation uses roughly x30 of the size of key-value, but this would put us in still reasonable amount of 4 * 30 * 3072 = 360kB.

**Comment by gmarek** [comment]
@k8s-bot ok to test

**Comment by gmarek** [comment]
panic: runtime error: index out of range

```
goroutine 1 [running]:
k8s.io/kubernetes/pkg/controller/node.(*NodeController).Run(0xc820294700, 0x2540be400)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/node/nodecontroller.go:219 +0x2be
k8s.io/kubernetes/cmd/kube-controller-manager/app.(*CMServer).Run(0xc820240d20, 0xc82029e520, 0x0, 0x2, 0x0, 0x0)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-controller-manager/app/controllermanager.go:310 +0x81e
main.main()
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-controller-manager/controller-manager.go:51 +0xe8
```

**Comment by WeixuZhuang** [comment]
Hi, could u share the logs? It seems that it encounter an out of index error when calling 
cidrIP := nc.clusterCIDR.IP.To4() and cidrIP[1] and cidrIP[2]. This code seats in the beginning of Run(). It seems we have not get the nc.clusterCIDR at that time. I will move this piece of code back to reconcileNodeCIDRs func where the original code generate availableCIDR. I hope this will resolve the issue. Thanks

**Comment by gmarek** [comment]
LGTM, but I'll apply the label after holiday, so I'll be able to monitor tests.

Thanks for helping with this.

**Comment by WeixuZhuang** [comment]
Thanks for your help @gmarek , have a nice holiday

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by gmarek** [comment]
`ERROR: Error fetching remote repo 'origin'`

@k8s-bot test this

**Comment by gmarek** [comment]
@k8s-bot unit test this

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by WeixuZhuang** [comment]
I just rebase my code onto the latest master and run unit test in my local machine. It pass the unit test and then I push out the changes. I don't why it fail the unit test here, any ideas?

**Comment by gmarek** [comment]
@k8s-bot unit test this

**Comment by gmarek** [comment]
I love this level of flakyness...
@k8s-bot unit test this

**Comment by gmarek** [comment]
@k8s-bot unit test this

**Comment by WeixuZhuang** [comment]
Finally, it passed the unit test....

Thank you so much @gmarek ^_^

**Comment by gmarek** [comment]
Flakyness rate is unbelievable...

**Comment by WeixuZhuang** [comment]
emm, never encountered this problem before...

could u mark it as LGTM and make it merged?

Thanks

**Comment by gmarek** [comment]
Oh, sorry - I didn't notice it was removed.

**Comment by spxtr** [comment]
@k8s-bot unit test this again

**Comment by gmarek** [comment]
@k8s-bot unit test this.

**Comment by therc** [comment]
I guess this was always busted in a sense, if you ran more than 256 nodes with a /16 CIDR, but for users within the limits, it wasn't broken. generateCIDRs would only generate as many candidates as nodes, all within the CIDR. PopAny() would thus be safe. But now, for my 10.65.0.0/16 example, you're just popping a random subnet out of 48896, which means you're almost always going to take one outside the CIDR.

**Comment by WeixuZhuang** [comment]
Hi @therc , I make that /24 assumption because in previous code, we also assume the cidr is /24. You could find it in line 264 to line 272 in old code.

I agree what it will give you more cidr than you should have when you pass /16 cidr into it. But our old code also did not handle the case of more than 256 nodes in one cluster(it always generate /24 cidr for cluster), if we decide to do so, it is not complicated to change my code and support that case. 

Thanks

**Comment by gmarek** [comment]
@WeixuZhuang - can you make a PR that'll make it safe for @therc case? We'll figure out how to deal with a bigger number of nodes later, but can you please make the generator respect CIDRs other than /24?

**Comment by therc** [comment]
The node CIDR is /24, yes. I'm not arguing about that, although one day it might be nice to support a larger mask. But the *cluster* CIDR can be a /16.

And although you are making the same assumption about a /8 (kind of! it's actually smaller... from e.g. 10.x.y.0 to 10.255.255.0), you are now guaranteed to break things, while the old code would only break if enough nodes were in the cluster. That's a big difference.

Something like this would work:
```
cidrSize, _ := nc.clusterCIDR.Mask.Size()
nc.maxCIDRs = int(math.Pow(2, (float64)(32-cidrSize-8)))
```

It would also be great if the code called IPNet.Contains() on clusterCIDR.IP and stopped from handing out a bogus CIDR outside of the cluster range.

**Comment by WeixuZhuang** [comment]
Hi @therc , I got you. Both my code and old code hold the assumption that the cluster CIDR is /8 and it may fail when it comes to /16 or something else. 

Also, I like your suggestion and I could do a quick fix by adding a verification func such as using IPNET.Contains(). Thanks for your suggestion. I will make a new PR by the end of today

**Comment by WeixuZhuang** [comment]
The code will be something like:
cidrSize, _ := nc.clusterCIDR.Mask.Size()
nc.maxCIDRs = int( math.Mod( (256-cidr[0])*2^24 + (256-cidr[1])*2^16 + (256-cidr[2])*2^8, 2^(32-cidrSize) ) / (2^8) )

where 2^k means math.Pow(2, k), I will reference the url in this conversation when I make a new PR.

Thanks so much

**Comment by WeixuZhuang** [comment]
https://github.com/kubernetes/kubernetes/pull/19580

This is the new PR, thanks
